content/blog/2019-10-23-divergence-optimizations/index.md

```diff
@@ -0,0 +1,231 @@
+| ------------- | ------------- | 
+| vadd  | c[i] = a[i] + b[i]  |
+| vsub  | c[i] = a[i] - b[i]  | 
+| vmul  | c[i] = a[i] * b[i]  | 
+| vdiv  | c[i] = a[i] / b[i]  | 
+| s2v  | c = (v0,v1,v2,v3)  | 
+| s2vb  | c = (v0,v0,v0,v0)  | 
+| v2s  | c = v[i]  | id |
+| gather  | c[0,1,2,3] = mem[a0,a1,a2,a3]  | 
+| scatter  | mem[a0,a1,a2,a3] = a[0,1,2,3]  | 
+| vload  | c[i] = mem[i]  | 
+| vstore  | mem[i] = a[i]  | 
+| vcmp  | pred = a[i] == b[i]  | 
+| vphi  | c[i] = pred ? a[i] : b[i]  |
+
+## Divergence Analysis
+
+Divergence analysis statically determines whether a vector instruction has redundant lanes of computation. In the following code, if `vec0` and `vec1` are vectors with the same value in each index, then the vector add will do the **exact** same work in each lane of the ALU. It would be much more efficient to do a single scalar (single-lane) `add` instruction instead.
+
+```python
+# initialize vectors
+  vec0: vector = (v0, v0, v0, v0);
+  vec1: vector = (v1, v1, v1, v1);
+
+# add vectors
+  vec2: vector = vadd vec0 vec1;
+
+```
+
+An instruction is assumed to be convergent (not divergent) by default. We traverse the dataflow graph forwards and mark an instruction as divergent if the following conditions are met. Our algorithm is based on the descriptions in [these](https://dl.acm.org/citation.cfm?id=3314902) [papers](https://ieeexplore.ieee.org/document/6113840).
+
+|     Condition    | Description |
+| ------------- | ------------------- |
+| Instruction is `s2v` | A different scalar value is loaded into each index of a vector register |
+| Instruction is `vload`  | Unknown values are loaded into each element of a vector register |
+| Any data dependency is divergent  | An incoming edge in the dataflow graph is already divergent due to one of the previous conditions |
+
+It's possible that during runtime some vectors marked as divergent might turn out to be convergent. For example, a `vload` may load in contiguous elements with the same values. However, there is no way to optimize for this case statically.
+
+## Divergence Optimizations
+
+### Instruction Swapping
+
+Once we know which instructions are divergent and which are not, we can optimize the code on an instruction-by-instruction basis. In the previous Bril example, every vector instruction is convergent. Therefore we can swap each vector instruction with a more energy-efficient scalar instruction. The optimization is shown below. 
+
+```python
+# initialize vectors -> initialize scalars
+  vec0_s: int = id v0;
+  vec1_s: int = id v1;
+
+# add vectors -> add scalars
+  vec2_s: int = add vec0_s vec1_s;
+
+```
+
+We implement a 'swap table' that matches a vector instruction with a functionally equivalent scalar instruction. An alternate design would be to annotate each original scalar instruction with a vector length and just change the vector length instead of doing a swap. Our swap table is given below along with a description of each instruction reproduced from above.
```

Comment:
```suggestion
We implement a "swap table" that matches a vector instruction with a functionally equivalent scalar instruction. An alternate design would be to annotate each original scalar instruction with a vector length and just change the vector length instead of doing a swap. Our swap table is given below along with a description of each instruction reproduced from above.
```